bgpd/bgp_dump.c

```diff
@@ -453,7 +451,7 @@ static int bgp_dump_interval_func(struct thread *t)
 {
 	struct bgp_dump *bgp_dump;
 	bgp_dump = THREAD_ARG(t);
-	bgp_dump->t_interval = NULL;
+	THREAD_OFF(bgp_dump->t_interval);
```

Comment:
the pattern we are supposed to use is when we create the thread of execution we pass in the address of the thread pointer.  in this case:

		thread_add_event(bm->master, bgp_dump_interval_func, bgp_dump,
				 0, &bgp_dump->t_interval);

when we get into bgp_dump_interval_func bgp_dump->t_interval is already null.  There is nothing to do here.  As such the above diff is wrong we do not need it at all


In addition we have this pattern:
		/* One-off dump: execute immediately, don't affect any scheduled
		 * dumps */
		bgp_dump->t_interval = NULL;
		thread_add_event(bm->master, bgp_dump_interval_func, bgp_dump,
				 0, &bgp_dump->t_interval);

Which just drops the bgp_dump->t_interval pointer if it was already set.  This creating a situation where we are could be calling the bgp_dump_interval_func function two times when we don't mean to.
the setting of bgp_dump->t_interval = NULL is incorrect in the above example here and should be removed or THREAD_OFF()

We cannot dump the thread pointers by setting them to null we *must* call the appropriate THREAD_OFF or _cancel routines to correctly clean up.  As that the thread code still has a pointer to the thread.  And that thread has pointers into our data structures, which can be freeed ( and this is the cause of the ospfv3 anvl crash Martin is reporting )

Reply:
part of this is the examples that folks find- sometimes it may be fine to NULL one of these pointers, but too many times folks don't really understand the exact details of the situation and just copy the pattern that they find somewhere, and that gets us into trouble. as Donald says, it costs very little to _avoid_ these open-coded patterns, and it does help to prevent problems.

Reply:
fixed.